src/compiler/sys.ts

```diff
@@ -298,6 +298,55 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories = createFileMap<FileWatcher>();
+                const watchedFiles = createFileMap<(fileName: string, removed?: boolean) => void>();
+                const currentDirectory = process.cwd();
+
+                return { addFile, removeFile };
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const path = toPath(fileName, currentDirectory, getCanonicalPath);
+                    const parentDirPath = toPath(ts.getDirectoryPath(fileName), currentDirectory, getCanonicalPath);
+
+                    if (!watchedDirectories.contains(parentDirPath)) {

```

Comment:
I'm not sure I follow. In the loose file scenario, do we ever watch folders, or just the loose files? (Or do we watch the folder also in case a `tsconfig.json` file is added)?

Per your below comment:

 > recursive directory watcher is used with `tsconfig.json` file without `files` array.

So then I assume we never add watchers to project root subdirs in this case? (Which is probably the most common case) What's the point if you are already watching the root dir recursively? Also, doesn't this contradict the concern above that:

> ...concern mentioned by Vlad regarding recursive watchers was that the user might receive too many irrelevant notifications from file activities not watched

If we are adding a recursive directory watcher on a `tsconfig.json` root (when a files array isn't specified), then seems either we have a problem based on this concern, or the concern is unfounded.